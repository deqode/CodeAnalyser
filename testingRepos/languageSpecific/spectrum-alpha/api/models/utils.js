// @flow
import { db } from 'shared/db';

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table: string, cb: Function) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true })
      .then(cursor => {
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
        return cursor;
      })
  );
};

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(row =>
      row
        .hasFields('lastSeen')
        .and(row('lastSeen').during(db.now().sub(current), db.now()))
    )
    .count()
    .default(0)
    .run();
};

// prettier-ignore
export const getGrowth = async (table: string, range: Timeframe, field: string, filter: ?mixed) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = () => {
  return db
    .table('coreMetrics')
    .orderBy(db.desc('date'))
    .limit(90)
    .orderBy('date')
    .run();
};
