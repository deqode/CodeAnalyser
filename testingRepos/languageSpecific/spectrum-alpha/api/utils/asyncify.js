// @flow
// Turn a callback-based listener into many async iterators without buffering
import { $$asyncIterator } from 'iterall';

type Listener = ((arg: any) => void) => Promise<any>;

const defaultOnError = (err: Error) => {
  throw new Error(err);
};

type Options = {|
  onError?: (err: Error) => void,
  filter?: (arg: any) => boolean,
|};

type Watcher = {
  filter?: (arg: any) => boolean,
  callback?: ({ done: boolean, value: any }) => void,
};

const asyncify = (listener: Listener) => {
  let watchers: Array<Watcher> = [];
  listener(value => {
    watchers.forEach(watcher => {
      if (watcher.callback && (!watcher.filter || watcher.filter(value))) {
        watcher.callback({ done: false, value });
      }
    });
  });

  return ({ filter, onError = defaultOnError }: Options = {}) => {
    let watcher: Watcher = { filter };
    let watching = true;
    const cleanup = () => {
      if (watching) {
        watching = false;
        watchers = watchers.filter(w => w !== watcher);
      }
    };
    try {
      return {
        next: () =>
          new Promise(resolve => {
            watcher.callback = resolve;
            watchers.push(watcher);
          }),
        return: () => {
          cleanup();
          return Promise.resolve({ done: true });
        },
        throw: err => {
          cleanup();
          onError(err);
          return Promise.reject(err);
        },
        [$$asyncIterator]() {
          return this;
        },
      };
    } catch (err) {
      onError(err);
    }
  };
};

export default asyncify;
