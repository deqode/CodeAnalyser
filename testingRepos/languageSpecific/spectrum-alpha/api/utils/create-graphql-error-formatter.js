// @flow
const debug = require('debug')('api:utils:error-formatter');
import { RateLimitError } from 'graphql-rate-limit';
import Raven from 'shared/raven';
import { IsUserError } from './UserError';
import type { GraphQLError } from 'graphql';

const queryRe = /\s*(query|mutation)[^{]*/;

const collectQueries = query => {
  if (!query) return 'No query';
  return query
    .split('\n')
    .map(line => {
      const m = line.match(queryRe);
      return m ? m[0].trim() : '';
    })
    .filter(line => !!line)
    .join('\n');
};

const errorPath = error => {
  if (!error.path) return '';
  return error.path
    .map((value, index) => {
      if (!index) return value;
      return typeof value === 'number' ? `[${value}]` : `.${value}`;
    })
    .join('');
};

const logGraphQLError = (req, error) => {
  debug('---GraphQL Error---');
  debug(error);
  error &&
    error.extensions &&
    error.extensions.exception &&
    debug(error.extensions.exception.stacktrace.join('\n'));
  if (req) {
    debug(collectQueries(req.body.query));
    debug('variables', JSON.stringify(req.body.variables || {}));
  }
  const path = errorPath(error);
  path && debug('path', path);
  debug('-------------------\n');
};

const createGraphQLErrorFormatter = (req?: express$Request) => (
  error: GraphQLError
) => {
  logGraphQLError(req, error);

  const err = error.originalError || error;
  const isUserError = err[IsUserError] || err instanceof RateLimitError;

  let sentryId = 'ID only generated in production';
  if (!isUserError || err instanceof RateLimitError) {
    if (process.env.NODE_ENV === 'production') {
      sentryId = Raven.captureException(
        error,
        req && Raven.parsers.parseRequest(req)
      );
    }
  }

  return {
    message: isUserError ? error.message : `Internal server error: ${sentryId}`,
    // Hide the stack trace in production mode
    stack:
      !process.env.NODE_ENV === 'production' ? error.stack.split('\n') : null,
  };
};

export default createGraphQLErrorFormatter;
