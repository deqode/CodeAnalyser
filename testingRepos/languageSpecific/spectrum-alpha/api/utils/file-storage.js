// @flow
require('now-env');

import type { FileUpload, EntityTypes } from 'shared/types';

const { FILE_STORAGE } = process.env;

const getUploadImageFn = () => {
  switch (FILE_STORAGE) {
    case 'local':
      return require('./file-system').uploadImage;
    case 's3':
    default:
      return require('./s3').uploadImage;
  }
};

const uploadImageFn = getUploadImageFn();

export const uploadImage = (
  file: FileUpload,
  entity: EntityTypes,
  id: string
): Promise<string> =>
  uploadImageFn(file, entity, id).catch(err => {
    throw new Error(err);
  });
