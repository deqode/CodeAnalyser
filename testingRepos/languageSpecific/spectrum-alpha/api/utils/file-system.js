// @flow
import uuidv4 from 'uuid/v4';
import fs from 'fs';

import type { FileUpload, EntityTypes } from 'shared/types';

const STORAGE_DIR = 'public/uploads';
const READ_WRITE_MODE = 0o777;

const dirExists = (path: string): Promise<boolean> =>
  new Promise(res => fs.access(path, fs.constants.F_OK, err => res(!!!err)));

const createUploadsDir = (path: string): Promise<void> =>
  new Promise(res =>
    fs.mkdir(path, READ_WRITE_MODE, err => {
      if (err) throw new Error(err);
      res();
    })
  );

export const uploadImage = async (
  file: FileUpload,
  entity: EntityTypes,
  id: string
): Promise<string> => {
  const result = await file;

  if (!(await dirExists(STORAGE_DIR))) {
    await createUploadsDir(STORAGE_DIR);
  }

  return new Promise(res => {
    const filePath = `${uuidv4()}-${entity}-${id}`;
    const { stream } = result;
    stream.pipe(fs.createWriteStream(`${STORAGE_DIR}/${filePath}`));
    stream.on('end', () => {
      res(encodeURI(`/uploads/${filePath}`));
    });
  });
};
