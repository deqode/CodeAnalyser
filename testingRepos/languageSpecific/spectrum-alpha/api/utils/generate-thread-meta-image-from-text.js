// @flow
// Generates a meta image which shows a title and a footer text on a nice Spectrum background.
import theme from 'shared/theme';
import { btoa } from 'b2a';
import { stringify } from 'query-string';

// NOTES(@mxstbr):
// Imgix has a useful ~text endpoint that allows us to add text to any image, but unfortunately that endpoint only allows us to add one piece of text to an imageâ€”but we need two!
// We work around this by generating two images that only contain the text of the title and the footer respectively, nothing else, and then using the blend option to blend the background image with the title image (which only has the title text) and used the mark option to add our footer text as a "watermark" to the footer.

const WIDTH = 1500;
const IMGIX_TEXT_ENDPOINT = 'https://assets.imgix.net/~text';

const TITLE_PARAMS = {
  w: WIDTH * 0.8, // 10% padding on each side
  h: 270, // Magic number, clips text after three lines
  txtsize: 56,
  txtlead: 16,
  txtfont: 'Helvetica,Bold',
  txtalign: 'left,middle',
  txtcolor: 'ffffff',
  // NOTE(@mxstbr): txtclip (i.e. ellipsis on overflowing text) only works with single-line text, which titles aren't, so this doesn't do anything rn
  txtclip: 'end,ellipsis',
};

const FOOTER_PARAMS = {
  h: 64,
  txtsize: 36,
  txtcolor: theme.brand.default.replace('#', ''),
  txtalign: 'right,middle',
  txtfont: 'Helvetica,Bold',
};

const BACKGROUND_URL = `https://spectrum.imgix.net/default_images/twitter-share-card.png`;

type GetMetaImageInput = {
  title: string,
  footer: string,
};

const generateImageFromText = ({
  title,
  footer,
}: GetMetaImageInput): ?string => {
  const base64title = btoa(title);
  const base64footer = btoa(footer);
  if (!base64title || !base64footer) return null;

  const titleUrl = `${IMGIX_TEXT_ENDPOINT}?${stringify(
    { ...TITLE_PARAMS, txt64: base64title.replace(/=/g, '') },
    { encode: false }
  )}`;
  const footerUrl = `${IMGIX_TEXT_ENDPOINT}?${stringify(
    { ...FOOTER_PARAMS, txt64: base64footer.replace(/=/g, '') },
    { encode: false }
  )}`;

  const base64titleurl = btoa(titleUrl);
  const base64footerurl = btoa(footerUrl);

  if (!base64titleurl || !base64footerurl) return null;

  const BACKGROUND_PARAMS = {
    w: WIDTH,
    bm: 'normal', // Blend the title normally, don't change opacity or color or anything, just overlay it
    by: 170, // Magic numbers that get the position right
    bx: 170,
    markalign: 'left,bottom', // Show the footer on the left side
    markpad: 24, // We overwrite the X pos, so the padding only applies on the y-axis
    markx: 140,
    blend64: btoa(titleUrl).replace(/=/g, ''),
    mark64: btoa(footerUrl).replace(/=/g, ''),
  };

  return `${BACKGROUND_URL}?${stringify(BACKGROUND_PARAMS, { encode: false })}`;
};

export default generateImageFromText;
