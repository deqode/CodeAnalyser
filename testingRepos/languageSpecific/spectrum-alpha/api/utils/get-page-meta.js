const { parse } = require('url');
const generateMetaInfo = require('../../shared/generate-meta-info');

// Don't even try if the path is /<value> any of these
// TODO: Longer, more complete deny list here
const PATH_DENY_LIST = ['robots.txt', 'home', 'messages', 'notifications'];

type Meta = {
  title: string,
  description: string,
  extra: string,
};

export default (
  url: string,
  request: (query: string) => Promise
): Promise<Meta> => {
  const { pathname = '' } = parse(url);
  const [, first, second, third] = pathname.split('/');
  if (third || first.length === 0) return Promise.resolve(generateMetaInfo());
  let promise;

  switch (first) {
    case 'explore': {
      promise = Promise.resolve(generateMetaInfo({ type: 'explore' }));
      break;
    }
    case 'thread': {
      /**
       * Thread
       */
      promise = request(/* GraphQL */ `
        {
          thread(id: "${second}") {
            content {
              title
              body
            }
            type
            channel {
              name
              isPrivate
            }
            community {
              profilePhoto
            }
          }
        }
      `).then(res => {
        const {
          thread: { type, content, channel },
        } = res.data;
        return generateMetaInfo({
          type: 'thread',
          data: {
            title: content.title,
            type,
            body: content.body,
            channelName: channel.name,
            privateChannel: channel.isPrivate,
            image: community.profilePhoto,
          },
        });
      });
      break;
    }
    case 'users': {
      /**
       * User
       */
      promise = request(/* GraphQL */ `
        {
          user(username: "${second}") {
            name
            username
            description
            profilePhoto
            # coverPhoto
          }
        }
      `).then(res => {
        const { user } = res.data;
        return generateMetaInfo({
          type: 'user',
          data: {
            username: user.username,
            description: user.description,
            name: user.name,
            image: user.profilePhoto,
          },
        });
      });
      break;
    }
    default: {
      const isDenyListed = PATH_DENY_LIST.includes(first);
      if (second && !isDenyListed) {
        /**
         * Channel
         */
        promise = request(/* GraphQL */ `
          {
            channel(channelSlug: "${second}", communitySlug: "${first}") {
              name
              description
              isPrivate
              community {
                name
                profilePhoto
              }
            }
          }
        `).then(res => {
          const {
            channel,
            channel: { community },
          } = res.data;
          return generateMetaInfo({
            type: 'channel',
            data: {
              name: channel.name,
              description: channel.description,
              communityName: community.name,
              private: channel.isPrivate,
              image: community.profilePhoto,
            },
          });
        });
      } else if (!isDenyListed) {
        /**
         * Community
         */
        promise = request(/* GraphQL */ `
          {
            community(slug: "${first}") {
              name
              description
              profilePhoto
            }
          }
        `).then(res => {
          const { community } = res.data;
          return generateMetaInfo({
            type: 'community',
            data: {
              name: community.name,
              description: community.description,
              image: community.profilePhoto,
            },
          });
        });
      }
      break;
    }
  }

  if (!promise) return Promise.resolve(generateMetaInfo());

  return promise.catch(err => {
    console.error(`⚠️ Failed to load metadata for ${url}! ⚠️`);
    console.error(err);
    return generateMetaInfo();
  });
};
