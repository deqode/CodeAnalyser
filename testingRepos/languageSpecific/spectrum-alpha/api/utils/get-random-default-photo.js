// @flow
const faker = require('faker');

// Helper function to get a random profile and cover photo
const PALETTE = [
  'blue',
  'blush',
  'cool',
  'green',
  'orange',
  'peach',
  'pink',
  'red',
  'teal',
  'violet',
];
export default () => {
  const color = faker.random.arrayElement(PALETTE);
  return {
    profilePhoto: `/default_images/profile-${color}.png`,
    coverPhoto: `/default_images/cover-${color}.svg`,
  };
};
