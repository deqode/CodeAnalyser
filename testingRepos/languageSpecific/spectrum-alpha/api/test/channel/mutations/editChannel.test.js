//@flow
import { request } from '../../utils';
import db from 'shared/testing/db';
import data from 'shared/testing/data';

const channel = data.channels[0];
// various permissions for Spectrum community
const owner = data.users.find(({ username }) => username === 'mxstbr');
const noPermissionUser = data.users.find(({ username }) => username === 'bryn');

afterEach(() => {
  return db
    .table('channels')
    .filter({ slug: channel.slug })
    .update({ ...channel })
    .run();
});

const variables = {
  input: {
    name: 'edited name',
    slug: 'edited-slug',
    description: 'edited description',
    isPrivate: false,
    channelId: channel.id,
  },
};

it('should edit a channel if user is owner', async () => {
  const query = /* GraphQL */ `
    mutation editChannel($input: EditChannelInput!) {
      editChannel (input: $input) {
        name
        slug
        description
        isPrivate
      }
    },
  `;

  const context = {
    user: owner,
  };

  expect.assertions(4);

  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
  expect(result.data.editChannel.name).toEqual(variables.input.name);
  expect(result.data.editChannel.slug).toEqual(variables.input.slug);
  expect(result.data.editChannel.description).toEqual(
    variables.input.description
  );
});

it('should not edit a channel if user is not owner', async () => {
  const query = /* GraphQL */ `
    mutation editChannel($input: EditChannelInput!) {
      editChannel (input: $input) {
        name
        slug
        description
        isPrivate
      }
    },
  `;

  const context = {
    user: noPermissionUser,
  };

  expect.assertions(1);

  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
});

it('should not edit a channel if user is not signed in', async () => {
  const query = /* GraphQL */ `
    mutation editChannel($input: EditChannelInput!) {
      editChannel (input: $input) {
        name
        slug
        description
        isPrivate
      }
    },
  `;

  expect.assertions(1);

  const result = await request(query, { variables });
  expect(result).toMatchSnapshot();
});
