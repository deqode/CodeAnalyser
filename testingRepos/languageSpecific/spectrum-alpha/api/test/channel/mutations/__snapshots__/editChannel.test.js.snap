// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should edit a channel if user is owner 1`] = `
Object {
  "data": Object {
    "editChannel": Object {
      "description": "edited description",
      "isPrivate": false,
      "name": "edited name",
      "slug": "edited-slug",
    },
  },
}
`;

exports[`should not edit a channel if user is not owner 1`] = `
Object {
  "data": Object {
    "editChannel": null,
  },
  "errors": Array [
    [GraphQLError: You don't have permission to make changes to this channel.],
  ],
}
`;

exports[`should not edit a channel if user is not signed in 1`] = `
Object {
  "data": Object {
    "editChannel": null,
  },
  "errors": Array [
    [GraphQLError: You must be signed in to make changes to this channel.],
  ],
}
`;
